### PR TITLE
fix(controller): Updates GetTaskAncestry to skip visited nod. Fixes #1907

### LINKED ISSUE #1907: GetTaskAncestry loop through steps that are already visited and cause performance issue
Checklist:

 * [ ] I've included the version.
 * [ ] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [ ] I've included the logs.
 
**What happened**:
When testing complicated DAG workflow (10*10 matrix with Cartesian product like dependencies) the workflow controller stuck at GetTaskAncestry, as it try to process every node even through they are already visited and take extremely long time  

**What you expected to happen**:
The visited ancestry node should not be processed again during the same GetTaskAncestry cycle for better performance 

**How to reproduce it (as minimally and precisely as possible)**:
workflow spec to reproduce 
https://github.com/StoneRelax/ArgoBenchmark/blob/master/argo-10by10.yaml

**Anything else we need to know?**:

**Environment**:
- Argo version:
```
$ argo version
```
- Kubernetes version :
```
$ kubectl version -o yaml
```

**Other debugging information (if applicable)**:
- workflow result:
```
argo get <workflowname>
```
- executor logs:
```
kubectl logs <failedpodname> -c init
kubectl logs <failedpodname> -c wait
```
- workflow-controller logs:
```
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

**Logs**

```
argo get <workflowname>
kubectl logs <failedpodname> -c init
kubectl logs <failedpodname> -c wait
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

### PR DESCRIPTION
**Pull request by StoneRelax**
Checklist:

* [b] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [c] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`. Why? for the release notes. 
* [ ] Optional. I've added My organization is added to the README.
* [ ] I've signed the CLA and required builds are green. 

 Fixes #1907

### CODE REVIEW

File: workflow/common/ancestry.go [hunk 1]
```diff
@@ -18,19 +18,23 @@ func GetTaskAncestry(ctx Context, taskName string, tasks []wfv1.DAGTask) []strin
...
 	var getAncestry func(s string)
 	getAncestry = func(currTask string) {
-		task := taskByName[currTask]
-		for _, depTask := range task.Dependencies {
-			getAncestry(depTask)
-		}
-		if currTask != taskName {
-			if _, ok := visited[currTask]; !ok {
-				visited[currTask] = getTimeFinished(ctx, currTask)
+		if !visitedFlag[currTask]{
```
Comment (alexec) [inline workflow/common/ancestry.go]: @jessesuen @sarabala1979 is there a good reason to revisit these?

Comment (StoneRelax) [reply workflow/common/ancestry.go]: is this being reviewed ? please let me know if I missed something

**Comment by StoneRelax** [comment]
finally resolved CLA issue. Is there any update on this plz?

**Comment by alexec** [comment]
Couple of things need doing:

1. CI stalled. If you `git merge master && git push` that should fix it.
2. I think tests need to be added.
3. I think @sarabala1979 should have the opportunity to review this.

**Comment by StoneRelax** [comment]
@simster7 thanks for reviewing. 
GetTaskAncestry is not expensive but it is recursively called, so if the DAG has too many tasks within nested dependencies it will result controller spending forever to do DAG validation like described in issues/1907 (wf spec to reproduce is provided). Since the task dependencies is unlikely to change during a single DAG validation cycle, it should skip visited node just like what do when ensuring DAG has no loop.
@alexec CI is now passing. For test case I think **workflow/common/ancestry_test.TestGetTaskAncestryForValidation** is ensuring the output of this method is the same with before. Let me know if you think more test is needed.

**Comment by simster7** [comment]
Thanks @StoneRelax!

**Pull request merged**